src/Microsoft.PowerShell.ConsoleGuiTools/ConsoleGui.cs

```diff
@@ -101,7 +105,71 @@ public void Start(ApplicationData applicationData)
                 columnWidthsSum--;
             }
 
-            win.Add(new Label(GetPaddedString(gridHeaders, columnWidths, offset + offset - 1)));
+            var filterLabel = new Label("Filter")
+            {
+                X = 2
+            };
+
+            // 7 here is to reserve space for the Apply button
+            var filterFieldWidth = usableWidth - filterLabel.Text.Length - 7;
+            var filterField = new TextField(".*")
+            {
+                X = Pos.Right(filterLabel) + 1,
+                Y = Pos.Top(filterLabel),
+                CanFocus = true,
+                Width = filterFieldWidth
+            };
+            filterField.Changed += FilterField_Changed;
+
+            var filterError = new Label(string.Empty)
+            {
+                X = Pos.Right(filterLabel) + 1,
+                Y = Pos.Top(filterLabel) + 1,
+                ColorScheme = Colors.Base,
+                Width = filterFieldWidth
+            };
+
+            var filterApplyButton = new Button("Apply")
+            {
+                // Pos.Right(filterField) returns 0
+                X = Pos.Right(filterLabel) + 2 + filterFieldWidth,
+                Y = Pos.Top(filterLabel),
+                Clicked = () =>
+                {
+                    FilterData(applicationData, filterField.Text.ToString(), columnWidths, offset, filterError);
+                    _list.SetSource(_itemList);
+                }
+            };
+
+            var header = new Label(GetPaddedString(gridHeaders, columnWidths, offset + offset - 1))
+            {
+                X = 0,
+                Y = 2
+            };
+
+            win.Add(filterLabel, filterField, filterError, filterApplyButton, header);
+            var headerLineText = new StringBuilder();
```

Comment:
please comment what this part is - the "adding underlines" part.

Reply:
ok